src/main/kotlin/org/jitsi/nlj/rtcp/KeyframeRequester.kt

```diff
@@ -31,35 +36,94 @@ import org.jitsi.rtp.rtcp.rtcpfb.payload_specific_fb.RtcpFbFirPacketBuilder
+                                    drop = true
+
+                                    requestKeyframe(pkt.mediaSourceSsrc)
+                                }
+                            }
+
+                        }
+                }
+            }
+        }
+
+        return if (drop) null else packetInfo
+    }
+
     // Map a SSRC to the timestamp (in ms) of when we last requested a keyframe for it
     private val keyframeRequests = mutableMapOf<Long, Long>()
     private var firCommandSequenceNumber: Int = 0
+    private val keyframeRequestsSyncRoot = Object()
+
     // Stats
     private var numKeyframesRequestedByBridge: Int = 0
     private var numKeyframeRequestsDropped: Int = 0
 
-    override fun observe(packetInfo: PacketInfo) {
-        //TODO: translation
-        //TODO: aggregation
+    private var hasPliSupport: Boolean = false
+    private var hasFirSupport: Boolean = true
+
+    private fun generateFirCommandSequenceNumber(mediaSsrc: Long, nowMs: Long): Int {
+        synchronized (keyframeRequestsSyncRoot) {
+            return if (nowMs - keyframeRequests.getOrDefault(mediaSsrc, 0) < WAIT_INTERVAL_MS) {
+                logger.cdebug { "Sent a keyframe request less than ${WAIT_INTERVAL_MS}ms ago for $mediaSsrc, " +
+                        "ignoring request" }
+                numKeyframeRequestsDropped++
+                -1
+            } else {
+                keyframeRequests[mediaSsrc] = nowMs
+                logger.cdebug { "Keyframe requester requesting keyframe with FIR for $mediaSsrc" }
+                numKeyframesRequestedByBridge++
+                firCommandSequenceNumber++
+            }
+        }
     }
 
     fun requestKeyframe(mediaSsrc: Long) {
-        //TODO(brian): for now hardcode to send an FIR
-        val now = System.currentTimeMillis()
-        if (now - keyframeRequests.getOrDefault(mediaSsrc, 0) < 100) {
-            logger.cdebug { "Sent a keyframe request less than 100ms ago for $mediaSsrc, ignoring request" }
-            numKeyframeRequestsDropped++
-        } else {
-            keyframeRequests[mediaSsrc] = now
-            val firPacket = RtcpFbFirPacketBuilder(
-                mediaSenderSsrc = mediaSsrc,
-                firCommandSeqNum = firCommandSequenceNumber++
-            ).build()
-            logger.cdebug { "Keyframe requester requesting keyframe with FIR for $mediaSsrc" }
-            numKeyframesRequestedByBridge++
-            processPacket(PacketInfo(firPacket))
+        val firCommandSeqNum = generateFirCommandSequenceNumber(mediaSsrc, System.currentTimeMillis())
+        if (firCommandSeqNum < 0) {
+            return
         }
+
+        val pkt = if (hasPliSupport) RtcpFbPliPacketBuilder(
+                mediaSenderSsrc = mediaSsrc
+        ).build() else RtcpFbFirPacketBuilder(
+                mediaSenderSsrc = mediaSsrc,
+                firCommandSeqNum = firCommandSeqNum
+        ).build()
+
+        processPacket(PacketInfo(pkt))
```

Comment:
The previous code could get away with calling `processPacket` because `transform` was a no-op, but this will loop back to this Node. We can use `next()` instead